### PR TITLE
[WIP] Add avl tests

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Adding AVL tests here.

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
@neithernut 

I experience a SEGFAULT here because of an unchecked ptr accessing in `regen_metadata()` called from  `insert_element_into_tree()` in `avl_add()`:

Backtrace (gdb):

```
#0  regen_metadata (node=0x604ad0) at /home/m/dev/c/libreset/src/libreset/avl.c:402
#1  0x00007ffff79cfdd7 in insert_element_into_tree (el=0x604890, root=0x604870) at /home/m/dev/c/libreset/src/libreset/avl.c:233
#2  0x00007ffff79cfcce in avl_add (avl=0x604870, d=0x7fffffffe1d4, hash=1) at /home/m/dev/c/libreset/src/libreset/avl.c:185
#3  0x0000000000401c78 in test_avl_add_multiple (_i=0) at /home/m/dev/c/libreset/tests/avl/avl_tests.c:74
#4  0x00007ffff7bd66ab in ?? () from /usr/lib/libcheck.so.0
#5  0x00007ffff7bd69f7 in srunner_run () from /usr/lib/libcheck.so.0
#6  0x000000000040284e in main () at /home/m/dev/c/libreset/tests/tests.c:33
```

This happens for the test "test_avl_add_multiple". Can you suggest a fix?

---

More information on how to reproduce/debug:

Run the checks manually with `CK_FORK="no" <debugger of choice> tests/check_tests` and set breakpoints at the test after breaking there, set a breakpoint in `insert_element_into_tree` and go on from there.

**Comment by neithernut** [comment]
Mh... I'll have a look at it.

**Comment by neithernut** [comment]
Please adapt as soon as PR #106 is merged.

**Comment by matthiasbeyer** [comment]
Sure. Lets focus on the interface for today and the other stuff will follow.

**Comment by matthiasbeyer** [comment]
@neithernut Still failing because f bug mentioned above. I will create an issue on this.

File: tests/avl/avl_tests.c [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+}
+END_TEST
+
+START_TEST (test_avl_alloc_destroy) {
+    struct avl* avl = avl_alloc();
+
+    ck_assert(avl != NULL);
+    ck_assert(avl->root == NULL);
+
+    ck_assert_int_eq(1, avl_destroy(avl, NULL));
```
Comment (neithernut) [inline tests/avl/avl_tests.c:23]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+}
+END_TEST
+
+START_TEST (test_avl_add) {
+    struct avl* avl = avl_alloc();
+
+    int data = 1;
+    rs_hash hash = 1;
+
+    struct avl_el* element = avl_add(avl, hash, &data, NULL);
```
Comment (neithernut) [inline tests/avl/avl_tests.c:33]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    ck_assert_int_eq(element->node_cnt, 0);
+    ck_assert_int_eq(element->hash, hash);
+
+    ck_assert(element->ll.head != NULL);
+    ck_assert(element->ll.head->next == NULL);
+    ck_assert(element->ll.head->data != NULL);
+    ck_assert(*(int*) element->ll.head->data == data);
+
+    free(element);
+    free(avl);
```
Comment (neithernut) [inline tests/avl/avl_tests.c:48]: You should call `avl_destroy()`, too.

File: tests/avl/avl_tests.c [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+}
+END_TEST
+
+START_TEST (test_avl_add_destroy) {
+    struct avl* avl = avl_alloc();
+
+    int data = 1;
+    rs_hash hash = 1;
+
+    avl_add(avl, hash, &data, NULL);
```
Comment (neithernut) [inline tests/avl/avl_tests.c:58]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+    ck_assert_int_eq(1, avl_destroy(avl, NULL));
```
Comment (neithernut) [inline tests/avl/avl_tests.c:60]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    struct avl* avl = avl_alloc();
+
+    int data[] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+    rs_hash hash[] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+
+    struct avl_el* elements[10];
+
+    int i;
+    for (i = 0; i < 10; i++) {
+        struct avl_el* element = avl_add(avl, hash[i], &data[i], NULL);
```
Comment (neithernut) [inline tests/avl/avl_tests.c:74]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+    int data[] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+    rs_hash hash = 0;
+    struct avl_el* element = NULL;
+    struct avl_el* cmp = NULL;
+
+    int i;
+    for (i = 0; i < 10; i++) {
+        if (element == NULL) {
+            element = avl_add(avl, hash, &data[i], NULL);
```
Comment (neithernut) [inline tests/avl/avl_tests.c:122]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        } else {
+            cmp = avl_add(avl, hash, &data[i], NULL);
```
Comment (neithernut) [inline tests/avl/avl_tests.c:124]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+            ck_assert(cmp == element); /* collisions */
+        }
+    }
+
+    ck_assert(avl->root != NULL);
+    ck_assert(avl->root->l == NULL);
+    ck_assert(avl->root->r == NULL);
+
+    ck_assert_int_eq(1, avl_destroy(avl, NULL));
```
Comment (neithernut) [inline tests/avl/avl_tests.c:133]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    struct avl* avl = avl_alloc();
+    int data        = 1;
+    rs_hash hash    = 1;
+    struct avl_el* found;
+
+    avl_add(avl, hash, &data, NULL);
+    found = avl_find(avl, hash, NULL, NULL);
+
+    ck_assert(found == avl->root);
+    ck_assert_int_eq(1, avl_destroy(avl, NULL));
```
Comment (neithernut) [inline tests/avl/avl_tests.c:147]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+END_TEST
+
+START_TEST (test_avl_find_single) {
+    struct avl* avl = avl_alloc();
+    int data        = 1;
+    rs_hash hash    = 1;
+    struct avl_el* found;
+
+    avl_add(avl, hash, &data, NULL);
+    found = avl_find(avl, hash, NULL, NULL);
```
Comment (neithernut) [inline tests/avl/avl_tests.c:144]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    struct avl* avl = avl_alloc();
+    int data[]      = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+    rs_hash hash[]  = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+    struct avl_el* found[10] = { NULL };
+
+    int i;
+    int j;
+
+    for (i = 0; i < 10; i++) {
+        avl_add(avl, hash[i], &data[i], NULL);
```
Comment (neithernut) [inline tests/avl/avl_tests.c:162]: The parameter `cfg` must not be NULL.

File: tests/avl/avl_tests.c [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    }
+
+    for (i = 0; i < 10; i++) {
+        found[i] = avl_find(avl, hash[i], NULL, NULL);
```
Comment (neithernut) [inline tests/avl/avl_tests.c:166]: The parameter `cfg` must not be NULL.

**Comment by neithernut** [comment]
Please rebase, once PR #119 is merged.

**Comment by matthiasbeyer** [comment]
@neithernut Thanks for solving the issue from #119 , there are others, but these should be only the passed `NULL` ptrs you mentioned.

---

Btw: You don't need to mention them everywhere, just leave a note on which function the stuff is wrong (if it is always the same error. Anyways I'll appreciate your work here! :+1:

**Comment by neithernut** [comment]
I assumed it would be easier to spot involcations you missed, because the corresponding diffs are not marked as outdated.

**Comment by matthiasbeyer** [comment]
That's of course a point.

**Comment by matthiasbeyer** [comment]
This is/was re-done in #156 . Closed.

**Pull request closed without merging**